src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,142 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPASITY = 10;
+    private static final int NUMBER_ONE = 1;
+    private int size;
+    private T[] array;
+
+    public ArrayList() {
+        array = (T[]) new Object[DEFAULT_CAPASITY];
+    }
+
     @Override
     public void add(T value) {
-
+        checkEnoughArraySize(size);
+        array[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+        checkEnoughArraySize(size);
 
+        if (index <= size) {
+            T[] secondArrayPart = getArrayCopy(array, index, array.length);
+            array[index] = value;
+
+            int startPointForIteration = index + NUMBER_ONE;
+            for (int i = startPointForIteration; i < secondArrayPart.length; i++) {
+                array[i] = secondArrayPart[i - startPointForIteration];
+            }
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("incorrect index");
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        checkEnoughArraySize(size + list.size());
 
+        for (int i = 0; i < list.size(); i++) {
+            array[i + size] = list.get(i);
+        }
+        size = size + list.size();
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexForGet(index);
+        T value = null;
+
+        for (int i = 0; i < array.length; i++) {
+            if (i == index) {
+                value = array[i];
+            }
+        }
+        return value;
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndexForGet(index);
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexForGet(index);
+        int startPointForCopySecondPart = index + 1;
+        T[] secondArrayPart = getArrayCopy(array, startPointForCopySecondPart, array.length);
+        T value = array[index];
+
+        for (int i = index; i < secondArrayPart.length; i++) {
+            array[i] = secondArrayPart[i - index];
+        }
```

Comment:
I think code will be less readable

Reply:
I don't think so, you made the method more complicated, instead of using just one copy method, you are getting a second part and saving it. For me, it's less readable than just an arrayCopyMethod